Global renku ini lock timeout

Most of our tests seem to be running pretty smoothly now, but there's still the occasional concurrency issue. One issue that I've been seeing more often is https://github.com/SwissDataScienceCenter/renku-python/pull/1925/checks?check_run_id=2020376036#step:7:126
```
____________ ERROR at setup of test_list_datasets_outside_a_context ____________
[gw1] linux -- Python 3.7.10 /opt/hostedtoolcache/Python/3.7.10/x64/bin/python

client = LocalClient(path=PosixPath('/tmp/pytest-of-runner/pytest-0/popen-gw1/test_list_datasets_outside_a_c0/964f2919bdddc5b9'... {'host': None, 'owner': None, 'name': None}}, _migration_type=<MigrationType.ALL: 7>, external_storage_requested=True)
directory_tree = PosixPath('/tmp/pytest-of-runner/pytest-0/popen-gw1/test_list_datasets_outside_a_c0/directory_tree')
/opt/hostedtoolcache/Python/3.7.10/x64/lib/python3.7/site-packages/pep8.py:110: FutureWarning: Possible nested set at position 1
  EXTRANEOUS_WHITESPACE_REGEX = re.compile(r'[[({] | []}),;:]')

    @pytest.fixture
    def client_with_datasets(client, directory_tree):
        """A client with datasets."""
        from renku.core.models.provenance.agents import Person
    
        person_1 = Person.from_string("P1 <p1@example.com> [IANA]")
        person_2 = Person.from_string("P2 <p2@example.com>")
    
        client.create_dataset(name="dataset-1", keywords=["dataset", "1"], creators=[person_1])
    
        with client.with_dataset("dataset-2", create=True) as dataset:
            dataset.keywords = ["dataset", "2"]
            dataset.creators = [person_1, person_2]
    
>           client.add_data_to_dataset(dataset=dataset, urls=[str(p) for p in directory_tree.glob("*")])

/home/runner/work/renku-python/renku-python/conftest.py:432: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
/home/runner/work/renku-python/renku-python/renku/core/management/datasets.py:558: in add_data_to_dataset
    lfs_paths = self.track_paths_in_storage(*files_to_commit)
/home/runner/work/renku-python/renku-python/renku/core/management/storage.py:57: in wrapper
    return fn(self, *args, **kwargs)
/home/runner/work/renku-python/renku-python/renku/core/management/storage.py:198: in track_paths_in_storage
    if file_size >= self.minimum_lfs_file_size:
/home/runner/work/renku-python/renku-python/renku/core/management/storage.py:136: in minimum_lfs_file_size
    size = self.get_value("renku", "lfs_threshold")
/home/runner/work/renku-python/renku-python/conftest.py:370: in mocked_get_value
    return original_get_value(self, section, key, config_filter=config_filter)
/home/runner/work/renku-python/renku-python/renku/core/management/config.py:142: in get_value
    config = self.load_config(config_filter=config_filter)
/home/runner/work/renku-python/renku-python/renku/core/management/config.py:96: in load_config
    with self.global_config_lock:
/opt/hostedtoolcache/Python/3.7.10/x64/lib/python3.7/site-packages/filelock.py:323: in __enter__
    self.acquire()
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = <filelock.UnixFileLock object at 0x7fdb7a00d310>, timeout = 0.0
poll_intervall = 0.05

    def acquire(self, timeout=None, poll_intervall=0.05):
        """
        Acquires the file lock or fails with a :exc:`Timeout` error.
    
        .. code-block:: python
    
            # You can use this method in the context manager (recommended)
            with lock.acquire():
                pass
    
            # Or use an equivalent try-finally construct:
            lock.acquire()
            try:
                pass
            finally:
                lock.release()
    
        :arg float timeout:
            The maximum time waited for the file lock.
            If ``timeout < 0``, there is no timeout and this method will
            block until the lock could be acquired.
            If ``timeout`` is None, the default :attr:`~timeout` is used.
    
        :arg float poll_intervall:
            We check once in *poll_intervall* seconds if we can acquire the
            file lock.
    
        :raises Timeout:
            if the lock could not be acquired in *timeout* seconds.
    
        .. versionchanged:: 2.0.0
    
            This method returns now a *proxy* object instead of *self*,
            so that it can be used in a with statement without side effects.
        """
        # Use the default timeout, if no timeout is provided.
        if timeout is None:
            timeout = self.timeout
    
        # Increment the number right at the beginning.
        # We can still undo it, if something fails.
        with self._thread_lock:
            self._lock_counter += 1
    
        lock_id = id(self)
        lock_filename = self._lock_file
        start_time = time.time()
        try:
            while True:
                with self._thread_lock:
                    if not self.is_locked:
                        logger().debug('Attempting to acquire lock %s on %s', lock_id, lock_filename)
                        self._acquire()
    
                if self.is_locked:
                    logger().info('Lock %s acquired on %s', lock_id, lock_filename)
                    break
                elif timeout >= 0 and time.time() - start_time > timeout:
                    logger().debug('Timeout on acquiring lock %s on %s', lock_id, lock_filename)
>                   raise Timeout(self._lock_file)
E                   filelock.Timeout: The file lock '/home/runner/.renku/renku.ini.lock' could not be acquired.
```

Which is us locking the global renku config (in `$HOME/.renku/renku.ini`). This could be solved by either mocking home directory to a temp directory for each test, which is probably safer for tests overall anyways, and/or by changing the way we lock things to have a read lock and a write lock, where the file being locked for read just prevents writes from happening, but not other reads, and being locked for write prevents reads from happening.

--- a/renku/cli/exception_handler.py
+++ b/renku/cli/exception_handler.py
@@ -59,6 +59,7 @@ from urllib.parse import urlencode
 
 import click
 import filelock
+import portalocker
 
 from renku.core.errors import MigrationRequired, ParameterError, ProjectNotSupported, RenkuException, UsageError
 
@@ -121,7 +122,7 @@ class IssueFromTraceback(RenkuExceptionsHandler):
             result = super().main(*args, **kwargs)
             return result
 
-        except filelock.Timeout:
+        except (filelock.Timeout, portalocker.LockException, portalocker.AlreadyLocked):
             click.echo(
                 (
                     click.style("Unable to acquire lock.\n", fg="red",) + "Hint: Please wait for another renku "

--- a/renku/core/management/config.py
+++ b/renku/core/management/config.py
@@ -23,7 +23,7 @@ from pathlib import Path
 
 import attr
 import click
-import filelock
+import portalocker
 from pkg_resources import resource_filename
 
 from renku.core.models.enums import ConfigFilter
@@ -70,15 +70,14 @@ class ConfigManagerMixin:
         return str(self.renku_path / self.CONFIG_NAME)
 
     @property
-    def global_config_lock(self):
-        """Create a user-level config lock."""
-        lock_file = f"{self.global_config_path}.lock"
+    def global_config_read_lock(self):
+        """Create a user-level config read lock."""
+        return portalocker.Lock(self.global_config_path, timeout=0, flags=portalocker.LOCK_SH | portalocker.LOCK_NB)
 
-        import logging
-
-        logging.getLogger("filelock").setLevel(logging.ERROR)
-
-        return filelock.FileLock(lock_file, timeout=0)
+    @property
+    def global_config_write_lock(self):
+        """Create a user-level config write lock."""
+        return portalocker.Lock(self.global_config_path, timeout=0, flags=portalocker.LOCK_EX | portalocker.LOCK_NB)
 
     def load_config(self, config_filter=ConfigFilter.ALL):
         """Loads local, global or both configuration object."""
@@ -93,7 +92,7 @@ class ConfigManagerMixin:
             config_files += [self.global_config_path, self.local_config_path]
 
         if config_filter != ConfigFilter.LOCAL_ONLY:
-            with self.global_config_lock:
+            with self.global_config_read_lock:
                 config.read(config_files)
         else:
             config.read(config_files)
@@ -118,9 +117,9 @@ class ConfigManagerMixin:
         filepath = self.global_config_path if global_only else self.local_config_path
 
         if global_only:
-            os.umask(0)
-            fd = os.open(filepath, os.O_CREAT | os.O_RDWR | os.O_TRUNC, 0o600)
-            with self.global_config_lock:
+            with self.global_config_write_lock:
+                os.umask(0)
+                fd = os.open(filepath, os.O_CREAT | os.O_RDWR | os.O_TRUNC, 0o600)
                 with open(fd, "w+") as file:
                     config.write(file)
         else:

--- a/setup.py
+++ b/setup.py
@@ -170,6 +170,7 @@ install_requires = [
     "pathspec>=0.7.0,<=0.8.0",
     "patool==1.12",
     "pluggy==0.13.1",
+    "portalocker>=2.2.1,<2.3",
     "psutil>=5.4.7,<=5.7.2",
     "pyasn1>=0.4.5,<=0.4.8",
     "pyjwt==2.0.0",

--- a/tests/cli/fixtures/cli_kg.py
+++ b/tests/cli/fixtures/cli_kg.py
@@ -54,6 +54,7 @@ def mock_kg():
             status_code = 404 if authorization == "Bearer renku-token" else 401
             return status_code, {"Content-Type": "application/json"}, ""
 
+        response.add_passthru("https://pypi.org/")
         response.add_callback(
             responses.GET, re.compile("http(s)*://renku.ch/knowledge-graph/datasets/.*"), callback=callback
         )

--- a/tests/cli/fixtures/cli_repository.py
+++ b/tests/cli/fixtures/cli_repository.py
@@ -64,15 +64,35 @@ def instance_path(renku_path, monkeypatch):
 def repository(tmpdir):
     """Yield a Renku repository."""
     from click.testing import CliRunner
+    from git.config import GitConfigParser, get_config_path
 
     from renku.cli import cli
 
     runner = CliRunner()
     with _isolated_filesystem(tmpdir, delete=True) as project_path:
-        result = runner.invoke(cli, ["init", ".", "--template-id", "python-minimal"], "\n", catch_exceptions=False)
-        assert 0 == result.exit_code
+        home = tmpdir.mkdir("user_home")
+        old_home = os.environ.get("HOME", "")
+        old_xdg_home = os.environ.get("XDG_CONFIG_HOME", "")
 
-        yield os.path.realpath(project_path)
+        try:
+            # NOTE: fake user home directory
+            os.environ["HOME"] = str(home)
+            os.environ["XDG_CONFIG_HOME"] = str(home)
+            with GitConfigParser(get_config_path("global"), read_only=False) as global_config:
+                global_config.set_value("user", "name", "Renku @ SDSC")
+                global_config.set_value("user", "email", "renku@datascience.ch")
+
+            result = runner.invoke(cli, ["init", ".", "--template-id", "python-minimal"], "\n", catch_exceptions=False)
+            assert 0 == result.exit_code
+
+            yield os.path.realpath(project_path)
+        finally:
+            os.environ["HOME"] = old_home
+            os.environ["XDG_CONFIG_HOME"] = old_xdg_home
+            try:
+                shutil.rmtree(home)
+            except OSError:  # noqa: B014
+                pass
 
 
 @pytest.fixture

--- a/tests/cli/test_config.py
+++ b/tests/cli/test_config.py
@@ -16,7 +16,11 @@
 # See the License for the specific language governing permissions and
 # limitations under the License.
 """Test ``config`` command."""
+import subprocess
+import sys
+
 import pytest
+from flaky import flaky
 
 from renku.cli import cli
 
@@ -168,3 +172,60 @@ def test_readonly_config(client, runner, project, config_key):
     result = runner.invoke(cli, ["config", "remove", config_key])
     assert 2 == result.exit_code
     assert f"Configuration {config_key} cannot be modified." in result.output
+
+
+def test_config_read_concurrency(runner, project, client, run):
+    """Test config can be read concurrently."""
+    result = runner.invoke(cli, ["config", "set", "test", "value"])
+    assert 0 == result.exit_code
+
+    command = [
+        "nice",  # NOTE: Set low priority to increase chance of concurrency issues happening
+        "-n",
+        "19",
+        sys.executable,
+        "-m",
+        "renku.cli",
+        "config",
+        "show",
+        "test",
+    ]
+
+    processes = []
+
+    for _ in range(20):
+        processes.append(subprocess.Popen(command, stdout=subprocess.PIPE, stderr=subprocess.PIPE))
+
+    assert all(p.wait() == 0 for p in processes)
+    assert all(p.stdout.read().decode("utf8") == "value\n" for p in processes)
+
+
+@flaky(max_runs=10, min_passes=1)
+def test_config_write_concurrency(runner, project, client, run):
+    """Test config can be read concurrently."""
+    result = runner.invoke(cli, ["config", "set", "test", "value"])
+    assert 0 == result.exit_code
+
+    write_command = [
+        "nice",  # NOTE: Set low priority to increase chance of concurrency issues happening
+        "-n",
+        "19",
+        sys.executable,
+        "-m",
+        "renku.cli",
+        "config",
+        "set",
+        "--global",
+        "test2",
+        "other_value",
+    ]
+
+    processes = []
+
+    for _ in range(20):
+        processes.append(subprocess.Popen(write_command, stdout=subprocess.PIPE, stderr=subprocess.PIPE))
+        processes.append(subprocess.Popen(write_command, stdout=subprocess.PIPE, stderr=subprocess.PIPE))
+
+    assert all(p.wait() == 0 for p in processes)
+    outputs = [p.communicate() for p in processes]
+    assert any("Unable to acquire lock" in o[0].decode("utf8") for o in outputs)

